### ISSUE TITLE
Out of Stock associated product but its color swatch appears as product image

### ISSUE THREAD

**Issue by ADDISON74** [OpenMage/magento-lts#1694]
Here are the steps to reproduce this issue.

1. Create a **Configurable Product** based on **Color** attribute which has in my case 2 defined values Blue and Grey.

2. In Images section add 3 images. The first image is used as Base/Small/Thumbnail, and the other two for swatches with these Labels: Blue-swatch, Grey-swatch.

3. Create **2 Associated Products** for each color. In Inventory section you should have Qty >0 and Stock Availability "In Stock".

Now load the Frontend.  You should see as options both colors as bellow. 

![1](https://user-images.githubusercontent.com/8360474/122677493-28629080-d1eb-11eb-942c-1d4cba105298.jpg)

4. Let's edit associated product for Grey color. Go to Inventory section and change Stock Availability to "Out of Stock". Reload the Frontend. Now option Grey disappeared which is correct but look at "MORE VIEWS" block bellow the product image. Now there are two thumbnails instead of one. The new addition is Grey-swatch image just see its title attribute bellow. 

![2](https://user-images.githubusercontent.com/8360474/122677902-0c5fee80-d1ed-11eb-979e-1fbabd69be83.jpg)

In conclusion, if one option is not available its swatch should be excluded completely from listing in Frontend. Unfortunately Magento treats its image defined with Label format **Color-Name-swatch** as a product image and this is definitely a bug.

Please note if you set the associated product to Disable its swatch disappears from MORE VIEWS block. See bellow.

![3](https://user-images.githubusercontent.com/8360474/122678041-a9bb2280-d1ed-11eb-8e16-ad3567d42d3a.jpg)

**Comment by elidrissidev** [comment]
I'm taking this issue.

**Comment by elidrissidev** [comment]
I have identified the problem and tried multiple fixes that work, but I'm not sure yet which one is appropriate and will not cause side effects.

In `Mage_ConfigurableSwatches_Helper_Mediafallback::attachProductChildrenAttributeMapping`, there is a loop that goes through all configurable children and checks if they have the `color` attribute and *are in stock* before adding them to the `$mapping` array. This array is then set as data on the product with key `child_attribute_label_mapping`.
https://github.com/OpenMage/magento-lts/blob/17128ba430b2d372da7901a22ee0f083b47d66e4/app/code/core/Mage/ConfigurableSwatches/Helper/Mediafallback.php#L109-L146

It is then getting used by `Mage_ConfigurableSwatches_Helper_Productimg::filterImageInGallery` to decide whether the image should be displayed in the gallery. Since the product is out of stock, it is not present in the mapping array and thus the condition at the end is returning `true`.
https://github.com/OpenMage/magento-lts/blob/17128ba430b2d372da7901a22ee0f083b47d66e4/app/code/core/Mage/ConfigurableSwatches/Helper/Productimg.php#L332-L358

The easiest solution that will work is to add a condition to check if the image label resembles a swatch like so:
```php
if (preg_match('/.+'.self::SWATCH_LABEL_SUFFIX.'$/', $image->getLabel())) {
    // This is a swatch image, do not show it
    return false;
}
```
But then this will not work if the image label is is named after the color name directly without the '-swatch' suffix
